zebra-state/src/service/finalized_state.rs

```diff
@@ -346,14 +353,40 @@ impl FinalizedState {
                 }
             }
 
+            let sapling_root = sapling_note_commitment_tree.root();
+            let orchard_root = orchard_note_commitment_tree.root();
+
+            // Create the history tree if it's the Heartwood activation block.
+            let heartwood_height = NetworkUpgrade::Heartwood
+                .activation_height(self.network)
+                .expect("Heartwood height is known");
+            match height.cmp(&heartwood_height) {
+                std::cmp::Ordering::Less => assert!(history_tree.is_none()),
+                std::cmp::Ordering::Equal => {
+                    history_tree = Some(HistoryTree::from_block(
+                        self.network,
+                        block.clone(),
+                        &sapling_root,
+                        &orchard_root,
+                    )?);
```

Comment:
Can you open a ticket to add tests for history errors in the finalized state?

I think we'll need to make them cached state tests (and work out how to feed them corrupt blocks), or do significant refactoring, or wait until after NU5 testnet activation. So I'm not sure it's worth spending time on them right now.